tests/unit/bases/test_buildd.py

```diff
@@ -1023,6 +1059,79 @@ def test_ensure_config_compatible_empty_config_returns_none(fake_executor, mock_
     )
 
 
+def test_ensure_setup_completed(fake_executor, logs, mock_load):
+    """Verify the setup was completed by checking the instance config file."""
+    mock_load.return_value = InstanceConfiguration(setup=True)
+
+    base_config = buildd.BuilddBase(alias=buildd.BuilddBaseAlias.JAMMY)
+
+    assert (
+        base_config._ensure_setup_completed(executor=fake_executor, deadline=None)
+        is None
+    )
+
+    assert "Instance has already been setup." in logs.debug
+
+
+@pytest.mark.parametrize(
+    "error, details",
+    [
+        (
+            ValidationError("test-error", InstanceConfiguration),
+            "Failed to parse instance configuration file.",
+        ),
+        (FileNotFoundError, "Failed to find instance config file."),
+    ],
+)
+def test_ensure_setup_completed_load_error(error, details, fake_executor, mock_load):
+    """Raise an error when the instance config cannot be loaded."""
+    mock_load.side_effect = error
+
+    base_config = buildd.BuilddBase(alias=buildd.BuilddBaseAlias.JAMMY)
+
+    with pytest.raises(BaseConfigurationError) as raised:
+        base_config._ensure_setup_completed(executor=fake_executor, deadline=None)
+
+    assert raised.value.brief == "Instance is not properly setup."
+    assert raised.value.details == details
+    assert (
+        raised.value.resolution == "Clean instance and retry the requested operation."
```

Comment:
So if I understand correctly, if the instance is not properly setup the user has to call `<application> clean` themselves right? Is this how it worked in core18/20? Have we thought about doing that for the user?

Reply:
> So if I understand correctly, if the instance is not properly setup the user has to call <application> clean themselves right?
Correct.  

> Is this how it worked in core18/20?
`core18|20` auto-cleaned in this scenario.

When launching an instance, `craft-providers` will auto-clean or raise an exception, depending on the problem:
##### Reasons to autoclean
- wrong OS
- wrong LXD id mapping

##### Reasons to raise an Exception
- failure to install an apt package, snap
- failure to set hostname
- failure to setup network

I think you are right that this is more appropriate for an auto-clean rather than raising an Exception.

Reply:
As discussed offline, I am now auto-cleaning when the instance is not fully setup, similar to core18|20.